### ISSUE TITLE
PyCharm inspections

### ISSUE THREAD

**Issue by matthewlloyd** [charliermarsh/ruff#3040]
(Opening a new issue to continue the discussion from #2972 which will be closed by https://github.com/charliermarsh/ruff/pull/3022)

I went through all the PyCharm inspections, filtered those I deemed implementable in a non-type-aware linter, and categorized them by whether they have Ruff or PyLint counterparts, mostly by running Ruff and PyLint directly on the examples provided in the PyCharm Settings panel.

There were five left over, which might merit consideration for inclusion as RUF rules.

### PyCharm inspections not currently covered by Ruff or PyLint

* Class-specific decorator is used outside the class (@classmethod or @staticmethod)
  ```python
  @staticmethod
  def a(): pass
  ```
* Incorrect property definition (e.g. getter that does not return or yield a value)
  ```python
  class A:
      @property
      def field(): pass
  ```
* Invalid definition of 'typing.NamedTuple' (field with default followed by one or more fields without)
  ```python
  import typing

  class FullName(typing.NamedTuple):
      first: str
      last: str = ""
      middle: str
  ```
* Non-optimal list declaration
  ```python
  l = [1]
  l.append(2)
  ```
* Shadowing names from outer scopes
  ```python
  def outer(p):
    def inner(p):
        pass
  ```

### Covered by Ruff or PyLint

* Assignments to 'for' loop or 'with' statement parameter (see above)
  - PLW2901 redefined-loop-name
* Assignment can be replaced with augmented assignment (a = a + b becomes a += b; arguably less useful)
  - PLR6104 consider-using-augmented-assign 
* Dictionary contains duplicate keys ({'a': 1, 'a': 2})
  - PLW0109 duplicate-key
* First argument of the method is reassigned (e.g. reassignment to self)
  - PLW0642 self-cls-assignment
* __init__ method that returns a value
  - PLE101 return-in-init
* Method is not declared static (but does not use self)
  - PLR6301 no-self-use
  - PLW0211 bad-staticmethod-argument (not quite the same thing)
* Missed call to '__init__' of the super class
  - PLW0231 super-init-not-called
* Redeclared names without usages (def x(): pass, followed later by x = 2, no uses in between)
  - F811
* Redundant boolean variable check (if b == True)
  - PLC0121 singleton-comparison
* Redundant parentheses
  - UP034
* Statement has no effect (e.g. comprises solely a variable name)
  - PLW0104 pointless-statement
  - also B018 as above
* Too complex chained comparisons ((x >= 0 and x <= 1) becomes (0 <= x <= 1)).
  - PLR1716 chained-comparison
* Unnecessary backslash - backslashes in places where line continuation is implicit inside (), [], and {}.
  - PyLint does not detect this, but pycodestyle has E502: "The backslash is redundant between brackets" (not yet implemented in Ruff)

**Comment by charliermarsh** [comment]
Thank you for this, super thorough and much appreciated!

**Comment by pkulev** [comment]
I'm looking forward to see those checks in ruff!
I'll expand here some moments regarding F811 and similar check in pycharm.

In the following example all redeclarations were highlighted by pycharm, but ruff alerts only on imports and function/class definitions. Variable redeclaration without usage passes the check.

```python
import os
import sys
import os  # ruff alerts F811

def foo():
    pass

def foo():  # ruff alerts F811
    pass

class Foo:
    pass

class Foo:  # ruff alerts F811
    pass

class Bar:
    x = 10
    x = 20  # No alert

BAR = "bar"
BAR = "baz"  # No alert
```

It would be great enable checking for variables too.

**Comment by urob** [comment]
Apologies for slight abuse of this issue. Is there an up-to-date list of PyCharm inspections that are implemented in either ruff or ty? Now that PyCharm natively supports both, it would be nice to have all those gathered in one place so that one can disable the corresponding PyCharm inspections.

(Ideally, PyCharm would automatically do this when ruff/ty are enabled, based on the actual ruff rules active for a given project.)

**Comment by ntBre** [comment]
I'm not aware of such a list, but if you find or make one, feel free to share it here!

**Comment by avalanchy** [comment]
@ntBre that would be the list https://www.jetbrains.com/help/inspectopedia/Python.html

All of them have an ID, for example Inspection ID: PyAttributeOutsideInitInspection

**Comment by pallgeuer** [comment]
Any update on getting the originally mentioned inspections into Ruff?
